test/cpp/ext/otel/otel_test_library.cc

```diff
@@ -0,0 +1,132 @@
+//
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include "test/cpp/ext/otel/otel_test_library.h"
+
+#include "absl/functional/any_invocable.h"
+#include "api/include/opentelemetry/metrics/provider.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "opentelemetry/sdk/metrics/meter_provider.h"
+#include "opentelemetry/sdk/metrics/metric_reader.h"
+
+#include <grpcpp/grpcpp.h>
+
+#include "src/core/lib/channel/call_tracer.h"
+#include "src/core/lib/config/core_configuration.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/end2end/test_service_impl.h"
+
+namespace grpc {
+namespace testing {
+
+void OTelPluginEnd2EndTest::SetUp(
+    const absl::flat_hash_set<absl::string_view>& metric_names,
+    opentelemetry::sdk::resource::Resource resource,
+    std::unique_ptr<grpc::internal::LabelsInjector> labels_injector,
+    bool global_meter_provider) {
+  // We are resetting the MeterProvider and OpenTelemetry plugin at the start
+  // of each test to avoid test results from one test carrying over to another
+  // test. (Some measurements can get arbitrarily delayed.)
+  auto meter_provider =
+      std::make_shared<opentelemetry::sdk::metrics::MeterProvider>(
+          std::make_unique<opentelemetry::sdk::metrics::ViewRegistry>(),
+          std::move(resource));
+  reader_.reset(new grpc::testing::MockMetricReader);
+  meter_provider->AddMetricReader(reader_);
+  grpc_core::CoreConfiguration::Reset();
+  grpc::internal::OpenTelemetryPluginBuilder ot_builder;
+  ot_builder.EnableMetrics(metric_names);
+  if (global_meter_provider) {
+    opentelemetry::metrics::Provider::SetMeterProvider(
+        opentelemetry::nostd::shared_ptr<opentelemetry::metrics::MeterProvider>(
+            std::move(meter_provider)));
+  } else {
+    ot_builder.SetMeterProvider(std::move(meter_provider));
+  }
+  ot_builder.SetLabelsInjector(std::move(labels_injector));
+  ot_builder.BuildAndRegisterGlobal();
+  grpc_init();
```

Comment:
Why do we need to explicitly call `grpc_init()` and `grpc_shutdown()` in these tests?